platform-sdk/swirlds-state-impl/src/main/java/com/swirlds/state/merkle/MerkleStateRoot.java

```diff
@@ -293,652 +131,199 @@ public T copy() {
-            final ReadableQueueState<E> instance = (ReadableQueueState<E>) queueInstances.get(stateKey);
-            if (instance != null) {
-                return instance;
-            }
+    // Migration things
 
-            final var md = stateMetadata.get(stateKey);
-            if (md == null || !md.stateDefinition().queue()) {
-                throw new IllegalArgumentException("Unknown queue state key '" + stateKey + "'");
-            }
+    // TODO: decide on method from VirtualMapMigration
+    @Override
+    public MerkleNode migrate(@NonNull final Configuration configuration, int version) {
+        if (version < CURRENT_VERSION) {
+            AtomicLong totalMigratedObjects = new AtomicLong(0);
+            AtomicLong totalMigrationTimeMs = new AtomicLong(0);
+            AtomicLong totalValidationTimeMs = new AtomicLong(0);
+
+            // Config constants
+            // Threads which iterate over the given Virtual Map, perform some operation and write into its own output queue/buffer
+            final int THREAD_COUNT = 1;
+            final long MEGA_MAP_MAX_KEYS_HINT = 1_000_000_000;
+            final boolean VALIDATE_MIGRATION_FF = true;
+
+            final MerkleDbConfig merkleDbConfig = configuration.getConfigData(MerkleDbConfig.class);
+            final var tableConfig = new MerkleDbTableConfig(
+                    (short) 1,
+                    DigestType.SHA_384,
+                    MEGA_MAP_MAX_KEYS_HINT,
+                    merkleDbConfig.hashesRamToDiskThreshold());
+            final var virtualMapLabel = "VirtualMap";
+            final var dsBuilder = new MerkleDbDataSourceBuilder(tableConfig, configuration);
+            final var virtualMap = new VirtualMap(virtualMapLabel, dsBuilder, configuration);
+
+            logger.info(STARTUP.getMarker(), "Migrating all of the states (Singleton, KV and Queue) to the one Virtual Map...");
+
+            logger.info(STARTUP.getMarker(), "Migrating KV states to the one Virtual Map...");
+
+            IntStream.range(0, getNumberOfChildren())
+                    .mapToObj(this::getChild)
+                    .filter(child -> child instanceof VirtualMap)
+                    .map(child -> (VirtualMap) child)
+                    .forEach(virtualMapToMigrate -> {
+
+                        final var labelPair = decomposeLabel(virtualMapToMigrate.getLabel());
+                        final var serviceName = labelPair.key();
+                        final var stateKey = labelPair.value();
+                        final var stateIdBytes = getVirtualMapKey(serviceName, stateKey);
+
+                        // TODO: double check here if it is valid to use #putBytes
+                        // TODO: check possibilities for optimization
+                        InterruptableConsumer<Pair<Bytes, Bytes>> handler = pair ->
+                                virtualMap.putBytes(stateIdBytes.append(pair.key()), pair.value());
```

Comment:
I am not sure if `stateIdBytes.append(key)` is faster or slower than `getVirtualMapKey(serviceName, stateKey, key)`. Worth measuring

Reply:
Yes, I marked this place for performance inspection as it is worth trying and measuring options.